### PR TITLE
Some type inference fixes - math functions

### PR DESCRIPTION
**Pull request by ranjanan**
As pointed out in #92, the functions can be better designed to use Julia's type inference. This PR begins the fixing.

### CODE REVIEW

File: src/math.jl [hunk 1]
```diff
@@ -14,10 +14,10 @@ for (op,fn) in ((:+,:af_add), (:.+,:af_add), (:-,:af_sub), (:.-,:af_sub), (:.*,:
                 (:.<<, :af_bitshiftl),(:>>, :af_bitshiftr), (:.>>, :af_bitshiftr),
                 (:^, :af_pow), (:.^, :af_pow ))
 
-    @compat @eval function $op{T,S}(a::AFArray{T}, b::AFArray{S}; batched = true)
+    @compat @eval function $op{T,S, N1, N2}(a::AFArray{T,N1}, b::AFArray{S,N2}; batched = true)
         ptr = new_ptr()
         eval($(quot(fn)))(ptr, a, b, batched)
```
Comment (SimonDanisch) [inline src/math.jl]: by the way, what's up with the eval here?

**Comment by ranjanan** [comment]
Looks like we have to drop v0.4 support if we are to make these changes.

File: src/math.jl (hunk 1 above)
Comment (ranjanan) [inline src/math.jl]: Do you not need that `eval` to call the appropriate function? Is there a cleaner way to writing this, then? 😄

Comment (SimonDanisch) [inline src/math.jl]: doens't a simple `$(fn)(ptr, a, b, batched)` work?

**Comment by gaika** [comment]
Huge improvement in speed in my use case (10^4 to 10^5 elements) .Is it possible to just define an empty Base.@pure macro when on v0.4 to make it portable?

**Comment by ViralBShah** [comment]
FWIW, I am ok dropping 0.4 support and having this package move faster.

**Comment by ranjanan** [comment]
Merging for now, will add fixes to other functions in other PRs.

**Pull request merged**